README
=======================================================================
                Net::LDAPapi Module v3.0.6 for Perl5
		by Quanah Gibson-Mount @ Symas, Inc.
	based on version 2.00 by Quanah Gibson-Mount @ Stanford University
	based on version 1.50 by Howard Chu @ Symas Corporation
	based on version 1.43 by Clayton Donley, <donley@wwa.com>
=======================================================================

COPYRIGHT
=========
   Copyright (c) 2017-2109 Quanah Gibson-Mount. All rights reserved
   Copyright (c) 2007 Quanah Gibson-Mount. All rights reserved
   Copyright (c) 2007 Board of Trustees, Leland Stanford Jr. University
   Copyright (c) 2003 Howard Chu. All rights reserved.
   Copyright (c) 1998 Clayton Donley. All rights reserved.
   This program is free software; you can redistribute it and/or
   modify it under the same terms as Perl itself.


INTRODUCTION
============

This module acts as a Perl5 interface to the LDAP C Development Kits from
OpenLDAP and Mozilla.  The module itself is written completely
in C.  Full documentation is included, as are commented example scripts. 
The current version supports the LDAPv3 API as implemented in OpenLDAP.

REQUIREMENTS
============

This module requires Perl5, a C compiler, and the LDAP libraries and include
files from one of the following: 

OpenLDAP C SDK (support added by Symas):
    http://www.openldap.org
Mozilla LDAP C SDK
    http://wiki.mozilla.org/LDAP_C_SDK

A C Compiler is not required for the NT version if you download Perl5
and module binaries and the Mozilla SDK (you must copy the DLL into your
system directory).


PLATFORMS SUPPORTED
===================

This version should be easily compiled on various platforms. It has been tested
extensively under Solaris and Linux.

As of 3.00, the module has not been tested on Windows.

Please let me know if you compile it successfully on a new platform or have
any problems. For problems, please include the compilation errors. 

Support for the Mozilla SDK has not been tested with the v3 rewrites.  Testing
welcomed.


NEW PERL-OO INTERFACE
=====================

As of version 1.40, a new Perl-OO layer has been added to allow more
object oriented access into the API.  All of the original commands are
unchanged, but are no longer documented.  You are encouraged to use the
new OO style routines, which are documented in the man-page.

Version 1.42 and above support passing named arguments to all methods.


UNIX INSTALLATION
=================

   After extracting the archive, simply type the following:

      $ perl5 Makefile.PL
      $ make

   Edit the file test.pl to contain your LDAP server, BASEDN, and filter
string.  Next type:

      $ make test

   It should complete all of its tests with no problems.  To install the
module you will likely need to become root and type:

      # make install

   That's it.  You can now 'use Net::LDAPapi' and make LDAP calls within your
PERL5 scripts with ease.

NT SOURCE INSTALLATION
======================

   I use Microsoft Visual C++ 5.0.  I am not certain what procedures you will
need to follow with other compilers.

   C:\> perl Makefile.PL

   Answer the questions appropriately.  Supply the path to PERL.EXE and the
Mozilla library and include files from their SDK.

   C:\> nmake

   The module will now be built.

   c:\> edit test.pl

   Change the LDAP server name and other attributes as necessary.

   C:\> nmake test

   If it passes most of the tests, it probably works, so you'll want to
install.  For this, just type:

   C:\> nmake install

   You can now do a 'use Net::LDAPapi' from your NT Perl modules.  You will
also want be be sure to copy the nsldap32v11.dll file into your system32
folder.

 
NT BINARY INSTALLATION
======================

   First, obtain Perl 5.004 binaries for Windows NT.  You can get it at the
following URL:  http://www.perl.com/CPAN-local/authors/id/GSAR/.  Simply get
one of the files starting with 'perl5.004'.

   You will also need the Mozilla LDAP SDK mentioned previously.  Copy the
nsldap32v11.dll file from that archive into your system32 folder.

   Finally, unzip the archive into the C:\ directory if you installed
PERL in C:\PERL.  Otherwise, extract to the parent directory of your PERL
installation directory.


EXAMPLES
========

   There exists six examples in the 'examples' directory.  The current
examples are:

      o ldapwalk.pl - This script demonstrates the use of Asynchronous LDAP
         calls to return all the attributes and values for entries matching
         a filter specified on the command line.  Simply change the variables
         at the start of the file to match your environment.

      o ldapwalk2.pl - This is the same as ldapwalk.pl, only it shows how to
         replace the portion that retrieves results and packages them in
         the hash of hashes with the new get_all_entries function.

      o testwrite.pl - This script demonstrates using synchronous Add, Modify,
         and Delete calls to manage an LDAP server.  You need update access
         to an LDAP server to run this example.  Once again, simply change
         the variables at the top of the file.

      o www-ldap.pl - This script uses both this module and the CGI.pm
         module available on the Internet (and included in 5.004).  It
         allows users to authenticate and modify their own information.
         You can easily customize this program, or learn from it and
         write your own CGIs.  The BIND and WEB_AUTHENTICATE subroutines
         would be especially useful in incorporating LDAP authentication
         into your own CGI scripts, even if the script has nothing to
         do with LDAP otherwise.

      o web500.pl - This script is a complete Web->LDAP gateway.  It uses
         CGI.pm and this module.  It can be easily customized, supports
         searching, modifications, and even things like jpegphoto uploads
         and displaying.  The original design used frames, but switched
         to a WebPH style interface for the search part on the advice of
         Douglas Gray Stephens.

      o updatepw.pl - Updates a password entry for a user from Unix into
         the LDAP server.  You'll likely have to tinker with this program
         to get it to do exactly what you want, but it exists as an idea
         as to how you can do this.

      o ldap_mod_attr.pl - Matches a filter or UID and makes specified changes
         to each entry.  Contributed by Andrew J Cosgriff.

   In addition, the test.pl program in the top level directory shows some
of the basic synchronous search functionality.


LDAP C API SUPPORT
==================

   This module supports direct perl access to all C API calls with the
exception of filter generation calls (since you can do this better in Perl
anyway).

   While direct access is available, it is suggested that you use the
Perl-OO style interface if you have never used the C API.


FEEDBACK
========
   Any feedback should be directed to quanah.gibsonmount@gmail.com


BUGS
====

   The non-OO stuff should work well.  Please let me know if I've introduced
any bugs in the OO stuff or the changed examples.

   Please report bugs at github - https://github.com/quanah/net-ldapapi/issues/

--
Clayton Donley
Rolling Meadows, IL, USA
email: donley@wwa.com
web:   http://www.wwa.com/~donley
CPAN:  /authors/id/CDONLEY

Howard Chu
Chief Architect, Symas Corporation	http://www.symas.com
Core Team, OpenLDAP Project		http://www.openldap.org

Quanah Gibson-Mount
email: quanah.gibsonmount@gmail.com
CPAN: /by-authors/id/M/MI/MISHIKAL
Platform Architect
Synacor, Inc				http://www.zimbra.com http://synacor.com/
Core Team, OpenLDAP Project		http://www.openldap.org


Changes
Revision history for Perl5 Net::LDAPapi Module.
3.0.6 Mon Oct 07 10:03:00 PST 2019
	- Fix memory leaks (Issue #51)
3.0.5 Fri Dec 01 10:03:00 PST 2017
	- Fix cookie handling with MMR configuration
	- Fix ASN1 encoding of sync cookie
3.0.4 Mon Nov 30 19:00:00 PST 2015
	- Fix undef comparison
	- Misc variable initializations to quiet warnings
	- Fixed sasl mechanisms initializtion
	- Examples cleanup
	- LDAPv3 extended operation support
	- New developer mode test suite
	- Fixed quanah/net-ldapapi#3: ldap_set_rebind_proc XS being called with invalid arguments from set_rebind_proc
	- Fixed quanah/net-ldapapi#6: ldap_sasl_bind has wrong prototype in LDAPapi.xs
	- Fixed quanah/net-ldapapi#8: search_s() clobbers ATTRS parameter
	- Fixed quanah/net-ldapapi#11: result() blocking when called with output from rename()
	- Fixed quanah/net-ldapapi#20: ldap_result() doesn't honour passed timeout value
	- Fixed quanah/net-ldapapi#21: ldap_set_option(LDAP_OPT_TIMEOUT, 1) on OpenLDAP returns -1
	- Fixed quanah/net-ldapapi#28: Server control responses get eaten after a NULL character in the berval
	- Fixed quanah/net-ldapapi#30: ldap_search_ext() and ldap_search_ext_s() segfault when used with timeout
	- Fixed quanah/net-ldapapi#31: ldap_result() and ldap_url_search_st() timeout parameters have a granularity of 1 second
	- Fixed quanah/net-ldapapi#40: Server control requests get eaten after a NULL character in the berval
3.0.3 Wed Aug 20 12:23:00 PST 2008
	- Add Convert::ASN1 requirement.
	- Fix error code handling (CPAN bug#35910)
3.0.2 Wed Jan 02 12:23:00 PST 2008
	- Fix various variable initializations and scope issues
	- Fix initialization of sasl parm
	- Fix test unbind
	- Add ldap_perror function
3.0.1 Thu Aug 09 17:22:15 PDT 2007
	- New 3.0.1 release, fixes sasl headers
3.0.0 Thu Jul 12 12:13:00 PDT 2007
	- New 3.0 release, using the LDAP V3 api
2.01  Fri Feb 09 08:41:46 PST 2007
	- Fix library linking
2.00  Tue Feb 06 18:46:38 PST 2007
	- Stripped out ISODE, UMich bits
	- Updated to refer to Mozilla C SDK
1.43  Sat Jun 13 02:39:15 CDT 1998
	- Added replacement for strcasecmp that was more portable
	- Fixed a few memory leaks caused by changes in 1.42 and
	  tested these changes on both Netscape and UMICH SDKs.
	- Changed the test.pl to be more intuitive and no longer require
	  modifications.
	- Made a few cosmetic changes to Makefile.PL
1.42  Sat Apr 25 22:27:22 CDT 1998
	- Added named arguments to all methods
	- Got rid of a few old ber_free's that were causing problems
	- Fixed the sort functions (Thanks to Miguel Marques)
	- Added the ldap_url functions
	- I was destroying error messages before I printed them in a few
	  of the ldapwalk examples.  Whoops.  Fixed.
	- Fixed problem with Out of Memory errors on some platforms with
	  the get_all_values function.
	- Fixed a multitude of problems with rebinding
	- Resolved issue resulting in dumped core on modify where NULL
	  values were passed.
	- Much more extensive testing by the author against both Netscape
	  and University of Michigan libraries.  Many small corrections.
	- Added Makefile.PL options for Netscape LDAPv3 SDK.  No v3 calls
	  added yet.
1.40a Wed Jan 28 11:20:51 CST 1998
	- sort_entries and multisort_entries have been added with the single
	  arguments $attr and \@attr respectively.  Neither seem to work
	  properly on my system when sorting anything besides DN (attr =
	  non-null).  Please try this function on your system and let me
	  know if it you can get it to work.  If so, I will finish adding
	  this function with another argument to specify a perl based compare
          function for the sort.
	- Lots of changes from Hallvard B. Furuseth <h.b.furuseth@usit.uio.no>
          Including:
           o Got rid of extra checks in LDAPapi.pm where checks exist in
	      XS or Perl.
	   o Cleaned up ldap_<add/modify>[_s] with a few simple changes to
	      hash2mod() function.  Also fixed bug in that function that
	      causes problems when you supply an empty hash.
	   o Small typos and bug fixes in test.pl fixed.
	   o Error string is now retrieved when requested, rather than after
	      every call (What was I thinking when I did this?  Who knows....)
	   o Fixed some reverse logic in ldap_<set/get>_option in the XS file.
	- Changed memory related calls to use 'New' and 'Safefree' macros,
	  which seems to be what the perlguts document suggests.
	- Got rid of the ber_free define in ldap_compat.h when using UMICH.
	  This was causing core dumps on some platforms.
	- Added ldap_mod_attr.pl to the examples.  A wonderful script from
	  Andrew J Cosgriff <ajc@bing.wattle.id.au>.

1.40 Wed Jan 21 11:12:24 CST 1998
	- I never did include the new version of web500.pl that supports
          Add functionality.  It did exist, and I've included it with this
          release.
	- There was a bug when adding or modifying a record where the value
	  was set to be an integer.  The bug has probably existed since a very
	  old version of the module.  It has now been fixed.
	- Fix to internal rebind process for Netscape API users provided by
	  Rob Weltman <rweltman@netscape.com> of Netscape.

1.40beta Tue Jan 13 14:54:13 CST 1998
	- Wow, what hasn't changed in this version...
	- Removed the extra 'free' in ldap_get_dn that was submitted as
	  a fix to a memory leak.  The fix seems to give inconsistent results
	  on different platforms and APIs.
	- Added Perl-OO methods for virtually all API calls.  The C style
	  API calls still work, and will work in all future versions.
	  Note that I haven't added named arguments yet, so argument order
	  matters.  This should be in the next version.
	- Manpages completely rewritten.  Please send me any errors or typos.
	- All examples except the two web ones have been rewritten to run under
	  'use Strict'.
	- Test program and all examples except web500.pl have been rewritten
	  to use Perl-OO methods.  Might need to clean-up comments and code
	  after so many rewrites.
	- Minor bugfixes to web500.pl


examples/ldapwalk-dieter.pl
#!/usr/bin/perl
#
#  $Id: ldapwalk.pl,v 1.1 2007/07/01 20:14:40 dieter Exp dieter $
#  ldapwalk.pl - Walks through Records Matching a Given Filter
#  Author:  Clayton Donley, Motorola, <donley@cig.mot.com>
#
#  Demonstration of Synchronous Searching in PERL5.
#
#  Rather than printing attribute and values directly, they are
#  stored in a Hash, where further manipulation would be very simple.
#  The output could then be printed to a file or standard output, or
#  simply run through the modify or add commands.
#
#  Usage:  ldapwalk.pl FILTER
#  Example:  ldapwalk.pl "sn=Donley"
#

use strict;
use Net::LDAPapi;

#  Define these values

my $ldap_server = "localhost";
my $BASEDN = "dc=example, dc=com";
my $sizelimit = 100;            # Set to Maximum Number of Entries to Return
                               # Can set small to test error routines
my $deref = "search";

#  Various Variable Declarations...
my $ld;
my $dn;
my $attr;
my $ent;
my $ber;
my @vals;
my %record;
my $rc;
my $result;

#
#  Initialize Connection to LDAP Server

if (($ld = new Net::LDAPapi($ldap_server)) == -1)
{
   die "Connection Failed!";
}

#ldap_set_option(0,LDAP_OPT_DEBUG_LEVEL,-1);

#
#  Bind as NULL User to LDAP connection $ld

$ld->sasl_parms(-mech=>"DIGEST-MD5",-flags=>LDAP_SASL_AUTOMATIC);

if ($ld->bind_s("benchmark","xxx",LDAP_AUTH_SASL) != LDAP_SUCCESS)
# if ($ld->bind_s != LDAP_SUCCESS)
{
   $ld->unbind;
   die "bind: ", $ld->errstring, ": ", $ld->extramsg;
}
#
#  This will set the size limit to $sizelimit from above.  The command
#  is a Netscape addition, but I've programmed replacement versions for
#  other APIs.
$ld->set_option(LDAP_OPT_SIZELIMIT,$sizelimit);
# $ld->set_option(LDAP_OPT_DEREF,$deref);

#  This routine is COMPLETELY unnecessary in this application, since
#  the rebind procedure at the end of this program simply rebinds as
#  a NULL user.
#$ld->set_rebind_proc(&rebindproc);

#
#  Specify Search Filter and List of Attributes to Return

my $filter = $ARGV[0];
my @attrs = ("cn","mail","telephonenumber");

#
#  Perform Search
my $msgid = $ld->search($BASEDN,LDAP_SCOPE_ONELEVEL,$filter,\@attrs,0);

if ($msgid < 0)
{
   $ld->unbind;
   die "search: ", $ld->errstring, ": ", $ld->extramsg;
}

# Reset Number of Entries Counter
my $nentries = 0;

# Set no timeout.
my $timeout = -1;

#
#  Cycle Through Entries
while (($rc = $ld->result($msgid,0,$timeout)) == LDAP_RES_SEARCH_ENTRY)
{
  $nentries++;

  for ($ent = $ld->first_entry; $ent != 0; $ent = $ld->next_entry)
  {

#
#  Get Full DN

   if (($dn = $ld->get_dn) eq "")
   {
      $ld->unbind;
      die "get_dn: ", $ld->errstring, ": ", $ld->extramsg;
   }

#
#  Cycle Through Each Attribute

   for ($attr = $ld->first_attribute; $attr ne ""; $attr = $ld->next_attribute)
   {

#
#  Notice that we're using get_values_len.  This will retrieve binary
#  as well as text data.  You can change to get_values to only get text
#  data.
#
      @vals = $ld->get_values ($attr);
      $record{$dn}->{$attr} = [@vals];
   }
  }
  $ld->msgfree;

}
if ($rc == LDAP_RES_SEARCH_RESULT &&
     $ld->err != LDAP_SUCCESS)
{
   $ld->unbind;
   die "result: ", $ld->errstring, ": ", $ld->extramsg;
}

print "Found $nentries records\n";

$ld->unbind;

foreach $dn (keys %record)
{
   my $item;
   print "dn: $dn\n";
   foreach $attr (keys %{$record{$dn}})
   {
      for $item ( @{$record{$dn}{$attr}})
      {
         if ($attr =~ /binary/ )
         {
	    print "$attr: <binary>\n";
	 } elsif ($attr eq "jpegphoto") {
#
#  Notice how easy it is to take a binary attribute and dump it to a file
#  or such.  Gotta love PERL.
#
	    print "$attr: JpegPhoto (length: " . length($item). ")\n";
	    open (TEST,">$dn.jpg");
	    print TEST $item;
	    close (TEST);
         } else {
            print "$attr: $item\n";
         }
      }
   }
}

exit;

sub rebindproc
{

   return("","",LDAP_AUTH_SIMPLE);
}

